docs/diplomat-and-macros.md

```diff
@@ -0,0 +1,186 @@
+
+This is the exact same problem which caused us to decide to stop working on
+[#416](https://github.com/mozilla/uniffi-rs/pull/416) - the current world where the type universe
+is described externally doesn't have this problem - only the UDL file needs to be parsed when
+generating the foreign bindings - Rust code isn't considered. The application-services team has
+concluded that none of our non-trival use-cases for UniFFI could be described using macros,
+so supporting both mechanisms is pain for no gain.
+
+As noted in #416, `wasm-bindgen` has a similarly shaped problem, and solves it by having
+the Rust macro arrange for the resulting library to have an extra data section with the
+serialized "type universe" - foreign binding generation would then read this information from the
+already built binary. This sounds more complex than the UniFFI team has appetite for at
+the current time.
+
+## Is this a problem for users of diplomat? Will diplomat solve it?
+
+I couldn't find real examples using diplomat, so it's difficult to know if this
+is a problem in practice. UniFFI came from a world where we had Rust crates and
+a hand-written FFI that exposed types from all over the crate. If these tools
+had started with the limitations from the macro approach in mind, it's possible
+a different, acceptable design might have been made to work. Maybe duplicating
+some structs and supplying suitable `Into` implementations might make things workable?
+
+Diplomat comes from a very smart team. They may well come up with a novel solution, so
+UniFFI should track the progress of that project to see what we can gleefully steal
+in the future. As discussed below, a kind of "hybrid" approach might even be possible.
+
+# Looking forward
+
+Before looking forward, let's step back a little - both UniFFI and diplomat are solving the exact
+same use-cases, just using a different approach to defining the type universe.
+But if we ignore that, the tools take the same basic approach - they all build the
+type universe, then use the representation of this type universe to define both Rust
+and foreign bindings.
+
+The type universe described by diplomat is somewhat "leaner" than that described by UniFFI -
+Rust types are the first-class citizens in the universe. UniFFI defines an external type model -
+for example, there's a `Type` enum where, for example, `Type::Record(Record)` represents a
+Rust struct. In other words, diplomat's type world can not be divorced from Rust,
+whereas UniFFI's already is.
+
+That said though, there might be a future where merging or otherwise creating some
+interoperability between these type universes might make sense. You could imagine
+a world where you can use diplomat to describe your type universe, but use UniFFI's foreign
+generation code to generate the Kotlin bindings. Similarly, a world where you use UniFFI
+and UDL files to describe your type universe, but then use diplomat to generate
+the NodeJS bindings.
```

Comment:
Thinking a bit more seriously about this, I don't think that there's realistically much value in having a kind of two-way interoperability between UniFFI bits and Diplomat bits. It would require the projects to converge on a great many details of how the FFI layer works, and I don't think there are sufficiently many quick wins to fuel that kind of work. For example, it would almost certainly be quicker for Diplomat to gain a Kotlin backend by writing one from scratch than by trying to iterate the UniFFI Kotlin backend towards something that works for both.

However, what I could see happening in future is UniFFI becoming a kind of higher-level wrapper around Diplomat. I can imagine a Diplomat *backend* for UniFFI that converts a `.udl` file into a bridge module and then uses the Diplomat toolchain to generate bindings from it, keeping  some of the additional affordances/conveniences we've built for our specific use-cases (e.g. around megazording).

An interesting thought experiment, anyway 😁

Reply:
yeah, I did get a bit carried away - I've tried to tone this down a little in the new version.